internal/links/device.go

```diff
@@ -0,0 +1,198 @@
+/**
+# Copyright 2023 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package links
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	nvmlcore "github.com/NVIDIA/go-nvml/pkg/nvml"
+	"gitlab.com/nvidia/cloud-native/go-nvlib/pkg/nvlib/device"
+	"gitlab.com/nvidia/cloud-native/go-nvlib/pkg/nvml"
+)
+
+func asNvmlDevice(dev device.Device) (nvmlcore.Device, error) {
+	uuid, ret := dev.GetUUID()
+	if ret != nvml.SUCCESS {
+		return nvmlcore.Device{}, fmt.Errorf("falied to get device uuid: %v", ret)
+	}
+
+	device, ret2 := nvmlcore.DeviceGetHandleByUUID(uuid)
+	if ret2 != nvmlcore.SUCCESS {
+		return nvmlcore.Device{}, fmt.Errorf("failed to get device handle: %v", nvmlcore.ErrorString(ret2))
+	}
+	return device, ret
+}
+
+func GetP2PLink(dev1 device.Device, dev2 device.Device) (link P2PLinkType, err error) {
+	nvmlDev1, err := asNvmlDevice(dev1)
```

Comment:
This should not be required once the `GetTopologyCommonAncestor` function is included in the `nvml.Device` interface.